platform-sdk/swirlds-common/src/main/java/com/swirlds/common/config/sources/MappedConfigSource.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (C) 2022-2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.config.sources;
+
+import static com.swirlds.logging.LogMarker.CONFIG;
+
+import com.swirlds.base.ArgumentUtils;
+import com.swirlds.config.api.source.ConfigSource;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * A {@link ConfigSource} that can be used as a wrapper for other config sources, providing functionality to define
+ * mappings for given properties. This allows the same property value defined for one name in the wrapped
+ * {@link ConfigSource} to be accessed by both names.
+ * <p>
+ * For example, suppose you want to rename the "dbUrl" property to "general.databaseUrl". You can use a
+ * {@code MappingConfigSource} to create a mapping between "dbUrl" and "general.databaseUrl", so that you can use both
+ * names with the same value during the transition time.
+ * </p>
+ * <p>
+ * Note that multiple mappings defined for a property is not allowed and will throw an {@link IllegalArgumentException}
+ * at runtime.
+ * </p>
+ * <p>
+ * Note that adding a mapping to a property not defined in the original {@link ConfigSource} will throw an
+ * {@link IllegalArgumentException} at runtime.
+ * </p>
+ * <p>
+ * Note that the ordinal of this {@code ConfigSource} is taken from the original {@link ConfigSource} that was wrapped
+ * by this class.
+ * </p>
+ *
+ * @see ConfigSource
+ * @see ConfigMapping
+ */
+public class MappedConfigSource extends AbstractConfigSource {
+    private static final String PROPERTY_NOT_FOUND = "Property '{}' not found in original config source";
```

Comment:
I prefer to see the constant when it makes the code easier to read.
I'm more concerned that we use the clearer `.formatted` method on String, and avoid wrapping string constants unnecessarily.
Some inline strings force some very ugly wrapping behavior (because of the indentation acceleration of the "+8" standard), and we end up with string concatenation we can easily avoid.  A constant is always preferable to string concatenation, in my opinion.  If it's short and simple, however, inline is fine where it improves clarity.

For "logging" statements, however, inline strings are often clearer than declared constants (unless the string ends up line-wrapping).